### PR TITLE
Try using ZINB loss function

### PR DESCRIPTION
**Pull request by ajlee21**
This PR attempts to update the loss function used by the VAE to model the noise instead of the reconstruction loss in order to better account for the sparsity in the data.

This loss is based on this paper: https://www.nature.com/articles/s41467-018-07931-2
Code for their method can be found: https://github.com/theislab/dca/blob/master/dca/loss.py

The first change was to pull DCA's code into my ponyo code, which you can see in `cm_modules/helper_vae.py`. This was very complicated since my network implementation was different than theirs. Doesn't need a heavy review as most of this code was previously reviewed in ponyo

The second attempt was to use their DCA code as a base and add the KL loss to that. Most of the changes in this PR are modules from https://github.com/theislab/dca/blob/master/dca/loss.py which don't need a heavy review. At this point I have the VAE training on 1 epoch before it crashes and I'll need to look into what is causing this.

I will definitely need to refactor/clean up this code and dig into what is causing this error or move to pytorch. But I wantedt to submit what I have so far as I'll need to take a short break from this work and return to SOPHIE.

Any thoughts on what might be the issue in `1_train_vae.ipynb` would be helpful.

### CODE REVIEW

File: cm_modules/train.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    model = network.model
+    loss = network.loss
+
+    if output_dir is not None:
+        os.makedirs(output_dir, exist_ok=True)
+
+    if learning_rate is None:
+        optimizer = opt.__dict__[optimizer](clipvalue=clip_grad)
+    else:
+        optimizer = opt.__dict__[optimizer](lr=learning_rate, clipvalue=clip_grad)
```
Comment (ben-heil) [inline cm_modules/train.py]: I think the standard way to do this is to use the `getattr` builtin function in python. Your resulting code would look something like

``` python
OptimizerClass = getattr(opt, optimizer)
if learning_rate is None:
        optimizer = OptimizerClass(clipvalue=clip_grad)
    else:
        optimizer = OptimizerClass(lr=learning_rate, clipvalue=clip_grad)

Comment (ben-heil) [reply cm_modules/train.py]: Also, I don't _think_ that this is causing the error, but `optimizer` is both the parameter that creates the class and the variable the class is stored into. If for some reason this code is called multiple times that might lead to the error? It may be worth checking  by just renaming the `optimizer` parameter to `optimizer_name` or something

File: cm_modules/train.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    print("about to compiile")
+
+    # Instantiate an optimizer.
+    # if optimizer == "Adam":
+    #     optimizer = tf.keras.optimizers.Adam()
```
Comment (ben-heil) [inline cm_modules/train.py]: Can probably remove this block now that you have the more general version

File: cm_modules/train.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            )
+        )
+    )
+    model = network.model
+    loss = network.loss
+
+    if output_dir is not None:
+        os.makedirs(output_dir, exist_ok=True)
+
+    if learning_rate is None:
```
Comment (ben-heil) [inline cm_modules/train.py:62]: Do you still run into the error if you explicitly set the learning rate?

File: cm_modules/train.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        # output = np.array(adata.iloc[1:,1:])
+    print("output", output)
+    loss = model.fit(inputs, output,
+                     epochs=epochs,
+                     batch_size=batch_size,
+                     shuffle=True,
+                     callbacks=callbacks,
+                     validation_split=validation_split,
+                     verbose=verbose,
+                     **kwds)
```
Comment (ben-heil) [inline cm_modules/train.py:118]: Beyond what I've said so far I have no idea what is causing the error. It seems like the optimizer is getting deleted or recreated after the first epoch, but I don't know what would cause that. I also don't know how to debug things in tensorflow because none of the training logic is accessible

**Review by ben-heil — APPROVED** [review]
I added a couple theories about what might be causing the error, but I'm not super confident in any of them. Tensorflow/keras are mysterious to me

**Review by danich1 — APPROVED** [review]
LGTM after you update the issue Ben found.  
1. This [answer](https://github.com/keras-team/keras/issues/5427) here might be the implicit  cause for your issue. Basically your variable wasn't initialized when you ran the training loop and you need to do that first.
2. At this moment you can fix the bug with the solution above; however, this code looks like it is still using tensorflow 1. Since you are in early stages, I HIGHLY recommend either switching to tensorflow 2 or pytorch. It will make debugging/future work quite easier to handle if you switch.

**Comment by ajlee21** [comment]
Thank you both for your comments and suggestions. I tried to adopt DCA's environment, which uses TF v2, but I can check to see that I did indeed install the correct version. If this attempt doesn't work, I'll be moving to pytorch! For now, I seem to at least be making some progress on this, unlike the previous attempt where no matter what I tried I kept getting the same error.

File: cm_modules/train.py (hunk 3 above)
Comment (ajlee21) [reply cm_modules/train.py:62]: I'm not sure if I understand what you mean by "setting the learning rate explicitly". If I do something like this `keras.optimizer.Adam(learning_rate=0.001)` I do still get the same error.

**Comment by ajlee21** [comment]
It does look like I'm using tensorflow v2. Thank you both for taking the time to think about this a bit with me. Your thoughts are as much as I gathered as well based on the error, so I'll see if I can push this farther or it might actually be time for pytorch.

File: cm_modules/train.py (hunk 3 above)
Comment (ben-heil) [reply cm_modules/train.py:62]: Yeah that answers my question, thanks!
